internal/transports/p2p/network/address_book.go

```diff
@@ -0,0 +1,138 @@
+package network
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/bitcoin-sv/block-headers-service/internal/transports/p2p/peer"
+	"github.com/bitcoin-sv/block-headers-service/internal/wire"
+)
+
+// AddressBook represents a collection of known network addresses.
+type AddressBook struct {
+	banDuration  time.Duration
+	addrs        []*knownAddress // addrs is a slice containing known addresses
+	addrsLookup  map[string]int  // addrLookup is a map for fast lookup of addresses, maps address key to index in addrs slice
+	mu           sync.Mutex
+	addrFitlerFn func(*wire.NetAddress) bool
+}
+
+// NewAddressBook creates and initializes a new AddressBook instance.
+func NewAdressBook(banDuration time.Duration, acceptLocalAddresses bool) *AddressBook {
+	// Set the address filter function based on whether local addresses are accepted
+	addrFitlerFn := IsRoutable
+	if acceptLocalAddresses {
+		addrFitlerFn = IsRoutableWithLocal
+	}
+
+	const addressesInitCapacity = 500
+	return &AddressBook{
+		addrs:        make([]*knownAddress, 0, addressesInitCapacity),
+		addrsLookup:  make(map[string]int, addressesInitCapacity),
+		banDuration:  banDuration,
+		addrFitlerFn: addrFitlerFn,
+	}
+}
+
+// UpsertPeerAddr updates or adds a peer's address.
+func (a *AddressBook) UpsertPeerAddr(p *peer.Peer) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	pa := p.GetPeerAddr()
+	key, ka := a.internalFind(pa)
+
+	if ka != nil {
+		ka.peer = p
+	} else {
+		a.internalAddAddr(key, &knownAddress{addr: pa, peer: p})
+	}
+}
+
+// UpsertAddrs updates or adds multiple addresses.
+func (a *AddressBook) UpsertAddrs(address []*wire.NetAddress) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for _, addr := range address {
+		if !a.addrFitlerFn(addr) {
+			continue
+		}
+
+		key, ka := a.internalFind(addr)
+		// If the address is not found, add it to the AddressBook.
+		if ka == nil {
+			a.internalAddAddr(key, &knownAddress{addr: addr})
+		} else if addr.Timestamp.After(ka.addr.Timestamp) {
+			// Otherwise, update the timestamp if the new one is newer.
+			ka.addr.Timestamp = addr.Timestamp
+		}
+	}
+}
+
+// BanAddr bans a network address. Ignores address if doesn't exist in the AddressBook.
+func (a *AddressBook) BanAddr(addr *wire.NetAddress) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	_, ka := a.internalFind(addr)
+	if ka != nil {
+		now := time.Now()
+		ka.banTimestamp = &now
+	}
+}
+
+// GetRndUnusedAddr returns a randomly chosen unused network address.
+func (a *AddressBook) GetRndUnusedAddr(tries uint) *wire.NetAddress {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	alen := len(a.addrs)
+	for i := uint(0); i < tries; i++ {
+		// #nosec G404
+		ka := a.addrs[rand.Intn(alen)]
+		if ka.peer == nil {
+			if ka.isBanned(a.banDuration) {
+				continue
+			}
+			return ka.addr
+		}
+	}
+
+	// return nil if no suitable address is found
+	return nil
+}
```

Comment:
Maybe the `tries` param is not necessary here. Consider this implementation:
It requires more memory (can be optimized), but we are sure that some address will be picked (if possible)
```golang
func (a *AddressBook) GetRndUnusedAddr() *wire.NetAddress {
	a.mu.Lock()
	defer a.mu.Unlock()

	free := make([]*KnownAddress, 0, len(a.addrs))
	for _, ka := range a.addrs {
		if ka.peer == nil && !ka.isBanned(a.banDuration) {
			free = append(free, ka)
		}
	}

	if len(free) == 0 {
		// return nil if no free address is found
		return nil
	}

	return free[rand.Intn(len(free))].addr
}
```